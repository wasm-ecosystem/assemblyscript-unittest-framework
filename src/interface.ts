/**
 * This file define the interface of coverage information
 */
// input

import { Type } from "wasmparser";
import { ASUtil } from "@assemblyscript/loader";
import { relative } from "node:path";

// instrumented file information
export class InstrumentResult {
  baseName: string;
  constructor(baseName: string) {
    this.baseName = relative(process.cwd(), baseName).replaceAll(/\\/g, "/");
  }
  get sourceWasm() {
    return this.baseName.concat(".wasm");
  }
  get instrumentedWasm(): string {
    return this.baseName.concat(".instrumented.wasm");
  }
  get sourceMap(): string {
    return this.baseName.concat(".wasm.map");
  }
  get debugInfo(): string {
    return this.baseName.concat(".debugInfo.json");
  }
  get expectInfo(): string {
    return this.baseName.concat(".expectInfo.json");
  }
  get traceFile(): string {
    return this.baseName.concat(".trace");
  }
}

export type CodeSnippetIndex = number;
export type FunctionIndex = number;
export type LineIndex = number;
export type ColumnIndex = number;
export type FileIndex = number;
export type UncoveredBasicBlocks = Set<CodeSnippetIndex>;
export type UncoveredLines = Set<LineIndex>;

// input cov
export type BranchInfo = [CodeSnippetIndex, CodeSnippetIndex];
export type BranchInfos = BranchInfo[];

export type LineRange = [FileIndex, LineIndex, ColumnIndex][];
export type LineInfos = (LineRange | null)[];
export type LineInfoMap = Map<number, Set<LineIndex>>;

export interface CovDebugInfo {
  branchInfo: BranchInfos;
  index: FunctionIndex;
  lineInfo: LineInfos;
}

export interface CovInfo {
  branchInfo: BranchInfos;
  lineInfo: LineInfoMap;
}

export interface DebugInfo {
  debugInfos: Record<string, CovDebugInfo>;
  debugFiles: string[];
}

export type CovTrace = [FunctionIndex, CodeSnippetIndex];
export type TestCaseName = string;

export type ExpectInfoIndex = string;
export type AssertExpectValue = string;
export type AssertActualValue = string;
export type AssertMessage = [ExpectInfoIndex, AssertActualValue, AssertExpectValue];
export type AssertFailMessage = Record<TestCaseName, AssertMessage[]>;
export type CrashInfo = Set<TestCaseName>;
export type FailedLogMessages = Record<TestCaseName, string[]>;

export type FailedInfo = { hasCrash: boolean; assertMessages: string[]; logMessages: string[] };
export type FailedInfoMap = Map<TestCaseName, FailedInfo>;

export type ExpectInfoDebugLocation = string;
export type ExpectInfo = Record<ExpectInfoIndex, ExpectInfoDebugLocation>;

export interface IExecutionResult {
  fail: number;
  total: number;
  crashInfo: CrashInfo;
  failedInfo: AssertFailMessage;
  failedLogMessages: FailedLogMessages;
}

export interface ImportFunctionInfo {
  module: string;
  name: string;
  args: Type[];
  return: Type | undefined;
}

// output
export class Rate {
  used = 0;
  total = 0;
  getRate(): number {
    return this.total === 0 ? 100 : Math.round((this.used / this.total) * 10000) / 100;
  }
  toString(): string {
    return `${this.used}/${this.total}`;
  }
  static summarize(rates: Rate[]): Rate {
    return rates.reduce((prev, curr) => {
      prev.used += curr.used;
      prev.total += curr.total;
      return prev;
    }, new Rate());
  }
}
export class FileCoverageResult {
  constructor(public filename: string) {}
  statementCoverageRate: Rate = new Rate();
  branchCoverageRate: Rate = new Rate();
  functionCoverageRate: Rate = new Rate();
  lineCoverageRate: Rate = new Rate();
  sourceUsedCount: CodeCoverage[] = [];
  uncoveredlines: Set<number> = new Set();
}

export class FunctionCoverageResult {
  constructor(public functionName: string) {}
  branchCoverageRate: Rate = new Rate();
  uncoveredlines: UncoveredLines = new Set();
  lineRange: [number, number] = [Number.MAX_SAFE_INTEGER, Number.MIN_SAFE_INTEGER];
  /**
   * first means lineIndex;
   * second means usedCount;
   */
  sourceUsedCount = new Map<number, number>();

  /**
   * Now assemblyscrpt will compile foo<T>() to different function like foo<f64>() , foo<u32>() etc;
   * We need merge the generic function to foo() for coverage statistics
   */
  static mergeFromGeneric(nameWithoutGeneric: string, infos: FunctionCoverageResult[]): FunctionCoverageResult {
    const result = new FunctionCoverageResult(nameWithoutGeneric);
    result.lineRange = [
      Math.min(...infos.map((info) => info.lineRange[0])),
      Math.max(...infos.map((info) => info.lineRange[1])),
    ];
    result.branchCoverageRate = Rate.summarize(infos.map((info) => info.branchCoverageRate));
    for (const info of infos) {
      for (const line of info.uncoveredlines) result.uncoveredlines.add(line);
      for (const [lineIndex, count] of info.sourceUsedCount.entries()) {
        const srcLineUsedCount = result.sourceUsedCount.get(lineIndex);
        result.sourceUsedCount.set(lineIndex, srcLineUsedCount === undefined ? count : srcLineUsedCount + count);
      }
    }
    return result;
  }
}

export class CodeCoverage {
  static readonly default = -1;
  source: string;
  /**
   * Default means not a effect line;
   * 0 means not used but effect;
   * more than 0 means a used line;
   */
  usedCount: number = CodeCoverage.default;
  constructor(src: string) {
    this.source = src;
  }
}

export interface UnittestPackage {
  readonly testCodePaths: string[];
  readonly filterByName: (fullTestName: string) => boolean;
}

export interface TestNameInfo {
  testName: string;
  testFilePath: string;
}

export class ImportsArgument {
  module: WebAssembly.Module | null = null;
  instance: WebAssembly.Instance | null = null;
  exports: (ASUtil & Record<string, unknown>) | null = null;
  constructor(public framework: UnitTestFramework) {}
}

export type Imports = ((arg: ImportsArgument) => Record<string, unknown>) | null;

export interface TestOption {
  includes: string[];
  excludes: string[];
  testFiles?: string[];
  testNamePattern: string | null;
  collectCoverage: boolean;
  onlyFailures: boolean;

  flags: string;
  imports?: Imports;

  tempFolder: string;
  outputFolder: string;
  mode: OutputMode | OutputMode[];
  warnLimit?: number;
  errorLimit?: number;

  isolated: boolean;
}

export type OutputMode = "html" | "json" | "table";

export abstract class UnitTestFramework {
  /**
   * function to redirect log message to unittest framework
   * @param msg: message to log
   */
  abstract log(msg: string): void;
}

export const OrganizationName = "wasm-ecosystem";
export const Repository = "https://github.com/wasm-ecosystem/assemblyscript-unittest-framework";
