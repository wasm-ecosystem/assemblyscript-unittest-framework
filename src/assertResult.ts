import { promises } from "node:fs";
import { json2map } from "./utils/index.js";
import { FailedInfoMap, AssertMessage, ExpectInfo, IAssertResult } from "./interface.js";
import chalk from "chalk";

const readFile = promises.readFile;

export class AssertResult {
  fail = 0;
  total = 0;
  failedInfos: FailedInfoMap = new Map();

  async merge(result: IAssertResult, expectInfoFilePath: string) {
    this.fail += result.fail;
    this.total += result.total;
    if (result.fail > 0) {
      let expectInfo;
      try {
        const expectContent = await readFile(expectInfoFilePath, { encoding: "utf8" });
        expectInfo = json2map(JSON.parse(expectContent) as ExpectInfo);
        for (const [testcaseName, value] of json2map<AssertMessage[]>(result.failed_info)) {
          const errorMsgs: string[] = [];
          for (const msg of value) {
            const [index, actualValue, expectValue] = msg;
            const debugLocation = expectInfo.get(index);
            let errorMsg = `${debugLocation ?? ""}\tvalue: ${actualValue}\texpect: ${expectValue}`;
            if (errorMsg.length > 160) {
              errorMsg = `${debugLocation ?? ""}\nvalue: \n\t${actualValue}\nexpect: \n\t${expectValue}`;
            }
            errorMsgs.push(errorMsg);
          }
          this.failedInfos.set(testcaseName, {
            assertMessages: errorMsgs,
            logMessages: result.failedLogMessages[testcaseName],
          });
        }
      } catch (error) {
        if (error instanceof Error) {
          console.error(error.stack);
        }
        throw error;
      }
    }
  }

  print(log: (msg: string) => void): void {
    const render = (failed: number, total: number) =>
      (this.fail === 0 ? chalk.greenBright(total) : chalk.redBright(total - failed)) + "/" + this.total.toString();
    log(`\ntest case: ${render(this.fail, this.total)} (success/total)\n`);
    if (this.fail !== 0) {
      log(chalk.red("Error Message: "));
      for (const [testcaseName, { assertMessages, logMessages }] of this.failedInfos.entries()) {
        log(`\t${testcaseName}: `);
        for (const assertMessage of assertMessages) {
          log("\t\t" + chalk.yellow(assertMessage));
        }
        for (const logMessage of logMessages ?? []) {
          log(chalk.gray(logMessage));
        }
      }
    }
  }
}
