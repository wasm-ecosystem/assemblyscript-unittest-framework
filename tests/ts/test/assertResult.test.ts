import { join } from "node:path";
import { fileURLToPath, URL } from "node:url";
import { FailedInfoMap, IAssertResult } from "../../../src/interface.js";
import { AssertResult } from "../../../src/assertResult.js";
import chalk from "chalk";

const __dirname = fileURLToPath(new URL(".", import.meta.url));

test("no failed_info merge", async () => {
  const assertResult = new AssertResult();
  const testcaseA: IAssertResult = {
    fail: 0,
    total: 28,
    failed_info: {},
    failedLogMessages: {},
  };
  const expectInfoFIlePath = join(__dirname, "..", "fixture", "assertResultTest.expectInfo.json");
  await assertResult.merge(testcaseA, expectInfoFIlePath);
  expect(assertResult.fail).toEqual(0);
  expect(assertResult.total).toEqual(28);
  expect(assertResult.failedInfos).toEqual(new Map<string, string[]>());
});

test("equal failed", async () => {
  const assertResult = new AssertResult();
  const actualString = "A long sentence for testing errorMsg.length > 160 in assertResult.ts merge function";
  const expectString = "= A long sentence for testing errorMsg.length > 160 in assertResult.ts merge function ";
  const testcaseA: IAssertResult = {
    fail: 1,
    total: 28,
    failed_info: {
      A: [
        ["1", "100", "= 200"],
        ["3", "[10]", "= [1]"],
        ["12", "{ 1 : 1.5, 2 : 2.5 }", "= { 1: 1.5, 2 : 2.0 }"],
        ["11", actualString, expectString],
      ],
    },
    failedLogMessages: {
      A: ["log message 1", "log message 2", "log message 3"],
    },
  };
  const expectInfoFIlePath = join(__dirname, "..", "fixture", "assertResultTest.expectInfo.json");
  await assertResult.merge(testcaseA, expectInfoFIlePath);
  const expectFailedInfo: FailedInfoMap = new Map();
  expectFailedInfo.set("A", {
    assertMessages: [
      "tests/as/comparison.test.ts:10:20\tvalue: 100\texpect: = 200",
      "tests/as/comparison.test.ts:15:27\tvalue: [10]\texpect: = [1]",
      "tests/as/comparison.test.ts:59:22\tvalue: { 1 : 1.5, 2 : 2.5 }\texpect: = { 1: 1.5, 2 : 2.0 }",
      `tests/as/comparison.test.ts:48:47\nvalue: \n\t${actualString}\nexpect: \n\t${expectString}`,
    ],
    logMessages: ["log message 1", "log message 2", "log message 3"],
  });
  expect(assertResult.fail).toEqual(1);
  expect(assertResult.total).toEqual(28);
  expect(assertResult.failedInfos).toEqual(expectFailedInfo);
});

test("print", async () => {
  const assertResult = new AssertResult();
  const testcaseA: IAssertResult = {
    fail: 1,
    total: 28,
    failed_info: {
      A: [["1", "100", "= 200"]],
    },
    failedLogMessages: {
      A: ["log message 1", "log message 2", "log message 3"],
    },
  };
  const expectInfoFIlePath = join(__dirname, "..", "fixture", "assertResultTest.expectInfo.json");
  await assertResult.merge(testcaseA, expectInfoFIlePath);

  {
    const outputs: string[] = [];
    chalk.level = 0; // disable color
    assertResult.print((msg) => outputs.push(msg));
    expect(outputs.join("\n")).toMatchSnapshot();
  }
  {
    const outputs: string[] = [];
    chalk.level = 1; // force enable color
    assertResult.print((msg) => outputs.push(msg));
    expect(outputs.join("\n")).toMatchSnapshot();
  }
});
