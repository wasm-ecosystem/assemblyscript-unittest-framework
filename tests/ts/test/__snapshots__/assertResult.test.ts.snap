// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print 1`] = `
"
test case: 27/28 (success/total)

Error Message: 
	A: 
		tests/as/comparison.test.ts:10:20	value: 100	expect: = 200
log message 1
log message 2
log message 3"
`;

exports[`print 2`] = `
"
test case: [91m27[39m/28 (success/total)

[31mError Message: [39m
	A: 
		[33mtests/as/comparison.test.ts:10:20	value: 100	expect: = 200[39m
[90mlog message 1[39m
[90mlog message 2[39m
[90mlog message 3[39m"
`;
